README.md
# NAME

Prancer::Session::Store::Database

# SYNOPSIS

This module implements a session handler that stores sessions in a database. It
creates its own database connection, separate from any existing database
connection, to avoid any issues with transactions. It wraps all changes to the
database in transactions to ensure consistency.

This configuration expects a database table that looks like this:

    CREATE TABLE session (
        id CHAR(72) NOT NULL,
        application VARCHAR DEFALUT '' NOT NULL,
        timeout integer DEFAULT date_part('epoch'::text, now()) NOT NULL,
        data TEXT
    );

    CREATE UNIQUE INDEX session_uq ON sessions (id, application);
    CREATE INDEX session_timeout_ix ON sessions (timeout);

Additional columns may be added as desired but they will not be used by this
session handler.

To use this session handler, add this to your configuration file:

    session:
        store:
            driver: Prancer::Session::Store::Database::Driver::DriverName
            options:
                table: sessions
                database: test
                username: test
                password: test
                hostname: localhost
                port: 5432
                charset: utf8
                connection_check_threshold: 10
                expiration_timeout: 3600
                autopurge: 0
                autopurge_probability: 0.1
                application: foobar

# OPTIONS

- table

    The name of the table in your database to use to store sessions. This name may
    include a schema name, like `public.sessions`. Otherwise the default schema of
    the database user will be used. If this option is not provided then the default
    table name is `sessions`.

- database

    **REQUIRED** The name of the database to connect to. If using SQLite, this
    should be the path to the database file.

- username

    The username to use when connecting. If this option is not set the default is
    the user running the application server. If using SQLite then this will be
    ignored.

- password

    The password to use when connecting. If this option is not set the default is
    to connect with no password. If using SQLite then this will be ignored.

- hostname

    The host name of the database server. If this option is not set the default is
    to connect to localhost. If using SQLite then this will be ignored.

- port

    The port number on which the database server is listening. If this option is
    not set the default is to connect on the database's default port. If using
    SQLite then this will be ignored.

- charset

    The character set to connect to the database with. If this is set to "utf8"
    then the database connection will attempt to make UTF8 data Just Work if
    available.

- connection\_check\_threshold

    This sets the number of seconds that must elapse between calls to get a
    database handle before performing a check to ensure that a live database
    connection still exists. If the check for a live database connection fails then
    the session handler will attempt to reconnect. This handles cases where the
    database handle hasn't been used in a while and the underlying connection has
    gone away. If this is not set it will default to 30 seconds.

- timeout

    Tthis is the number of seconds a session should last in the database before it
    will be automatically purged. The default is to purge sessions after 1800
    seconds (30 minutes).

- autopurge

    This flag controls whether sessions will be automatically purged by Prancer.
    If set to 1, the default, then on 10% of requests to your application, Prancer
    will delete from the database any session that has timed out. If set to 0 then
    sessions will never be removed from the database. Note that this doesn't
    control whether sessions time out, only whether they get removed from the
    database.

- autopurge\_probability

    This is the probability that autopurge will run on any given request. By
    default, this value is 0.1, or 10%, meaning that 1 in every 10 requests will
    attempt to purge expired sessions. This can be set to "1" to purge on every
    session action or to something extremely small like 0.001 to purge very, very
    infrequently. Or you can export purging duties to another program entirely.

- application

    If multiple applications will be using the same session table then this option
    may be used to distinguish between them. This key will be used in the
    `application` column of the sessions table.

# COPYRIGHT

Copyright 2014 Paul Lockaby. All rights reserved.

This library is free software; you can redistribute it and/or modify it under
the same terms as Perl itself.

# SEE ALSO

- [Prancer](https://metacpan.org/pod/Prancer)
- [Prancer::Session](https://metacpan.org/pod/Prancer::Session)



Changes
Revision history for Perl extension Prancer-Session-Store-Database

1.00
	- Initial release.



Makefile.PL
#!/usr/bin/env perl

use strict;
use warnings FATAL => 'all';

use ExtUtils::MakeMaker;

WriteMakefile(
    'NAME'          => 'Prancer::Session::Store::Database',
    'AUTHOR'        => 'Paul Lockaby <plockaby AT cpan DOT org>',
    'VERSION_FROM'  => 'lib/Prancer/Session/Store/Database.pm',
    'PREREQ_PM'     => {
        'Prancer'                    => '1.00',
        'Plack'                      => '0.9968',  # same version required by Web::Simple
        'Plack::Middleware::Session' => '0.23',  # require secrets
        'DBI'                        => '0',
        'Try::Tiny'                  => '0',
        'perl'                       => '5.010',  # because "//" is used
    },
    'INSTALLDIRS'   => (($] < 5.011) ? 'perl' : 'site'),

    ((ExtUtils::MakeMaker->VERSION() lt '6.25') ?
        ('PL_FILES' => { })            : ()),
    ((ExtUtils::MakeMaker->VERSION() gt '6.30') ?
        ('LICENSE'  => 'perl')         : ()),
);



lib/Prancer/Session/Store/Database.pm
package Prancer::Session::Store::Database;

use strict;
use warnings FATAL => 'all';

use version;
our $VERSION = '1.00';

1;

=head1 NAME

Prancer::Session::Store::Database

=head1 SYNOPSIS

This module implements a session handler by storing sessions in a database. It
creates its own database connection, separate from any existing database
connection, avoid any issues with transactions. It wraps all changes to the
database in transactions to ensure consistency.

This configuration expects a database table that looks like this:

    CREATE TABLE session (
        id CHAR(72) NOT NULL,
        application VARCHAR DEFALUT '' NOT NULL,
        timeout integer DEFAULT date_part('epoch'::text, now()) NOT NULL,
        data TEXT
    );

    CREATE UNIQUE INDEX session_uq ON sessions (id, application);
    CREATE INDEX session_timeout_ix ON sessions (timeout);

Additional columns may be added as desired but they will not be used by this
session handler.

To use this session handler, add this to your configuration file:

    session:
        store:
            driver: Prancer::Session::Store::Database::Driver::DriverName
            options:
                table: sessions
                database: test
                username: test
                password: test
                hostname: localhost
                port: 5432
                charset: utf8
                connection_check_threshold: 10
                expiration_timeout: 3600
                autopurge: 0
                application: foobar

=head1 OPTIONS

=over 4

=item table

The name of the table in your database to use to store sessions. This name may
include a schema name. Otherwise the default schema of the user will be used.
If this option is not provided the default will be C<sessions>.

=item database

B<REQUIRED> The name of the database to connect to.

=item username

The username to use when connecting. If this option is not set the default is
the user running the application server.

=item password

The password to use when connectin. If this option is not set the default is to
connect with no password.

=item hostname

The host name of the database server. If this option is not set the default is
to connect to localhost.

=item port

The port number on which the database server is listening. If this option is
not set the default is to connect on the database's default port.

=item charset

The character set to connect to the database with. If this is set to "utf8"
then the database connection will attempt to make UTF8 data Just Work if
available.

=item connection_check_threshold

This sets the number of seconds that must elapse between calls to get a
database handle before performing a check to ensure that a database connection
still exists and will reconnect if one does not. This handles cases where the
database handle hasn't been used in a while and the underlying connection has
gone away. If this is not set it will default to 30 seconds.

=item timeout

This the number of seconds a session should last in the database before it will
be automatically purged. The default is to purge sessions after 1800 seconds.

=item autopurge

This flag controls whether sessions will be automatically purged by Prancer.
If set to 1, the default, then on 10% of requests to your application, Prancer
will delete from the database any session that has timed out. If set to 0 then
sessions will never be removed from the database. Note that this doesn't
control whether sessions time out, only whether they get removed from the
database.

=item application

If multiple applications will be using the same session table then this option
may be used to distinguish between them. This key will be included in the
application column of the database table.

=back

=cut


lib/Prancer/Session/Store/Database/Driver/Pg.pm
package Prancer::Session::Store::Database::Driver::Pg;

use strict;
use warnings FATAL => 'all';

use version;
our $VERSION = '1.00';

use Prancer::Session::Store::Database::Driver;
use parent qw(Prancer::Session::Store::Database::Driver);

use Try::Tiny;
use Carp;

# even though this *should* work automatically, it was not
our @CARP_NOT = qw(Prancer Try::Tiny);

sub new {
    my $class = shift;

    try {
        require DBD::Pg;
    } catch {
        my $error = (defined($_) ? $_ : "unknown");
        croak "could not initialize session handler: could not load DBD::Pg: ${error}";
    };

    my $self = bless($class->SUPER::new(@_), $class);
    my $database  = $self->{'_database'};
    my $username  = $self->{'_username'};
    my $password  = $self->{'_password'};
    my $hostname  = $self->{'_hostname'};
    my $port      = $self->{'_port'};
    my $charset   = $self->{'_charset'};
    my $table     = $self->{'_table'};

    my $dsn = "dbi:Pg:dbname=${database}";
    $dsn .= ";host=${hostname}" if defined($hostname);
    $dsn .= ";port=${port}" if defined($port);

    my $params = {
        'AutoCommit' => 0,
        'RaiseError' => 1,
        'PrintError' => 0,
    };
    if ($charset && $charset =~ /^utf8$/xi) {
        $params->{'pg_enable_utf8'} = 1;
    }

    $self->{'_dsn'} = [ $dsn, $username, $password, $params ];
    return $self;
}

1;


lib/Prancer/Session/Store/Database/Driver/MySQL.pm
package Prancer::Session::Store::Database::Driver::MySQL;

use strict;
use warnings FATAL => 'all';

use version;
our $VERSION = '1.00';

use Prancer::Session::Store::Database::Driver;
use parent qw(Prancer::Session::Store::Database::Driver);

use Try::Tiny;
use Carp;

# even though this *should* work automatically, it was not
our @CARP_NOT = qw(Prancer Try::Tiny);

sub new {
    my $class = shift;

    try {
        require DBD::mysql;
    } catch {
        my $error = (defined($_) ? $_ : "unknown");
        croak "could not initialize session handler: could not load DBD::mysql: ${error}";
    };

    my $self = bless($class->SUPER::new(@_), $class);
    my $database  = $self->{'_database'};
    my $username  = $self->{'_username'};
    my $password  = $self->{'_password'};
    my $hostname  = $self->{'_hostname'};
    my $port      = $self->{'_port'};
    my $charset   = $self->{'_charset'};
    my $table     = $self->{'_table'};

    my $dsn = "dbi:mysql:dbname=${database}";
    $dsn .= ";host=${hostname}" if defined($hostname);
    $dsn .= ";port=${port}" if defined($port);

    my $params = {
        'AutoCommit' => 0,
        'RaiseError' => 1,
        'PrintError' => 0,
    };
    if ($charset && $charset =~ /^utf8$/xi) {
        $params->{'mysql_enable_utf8'} = 1;
    }

    $self->{'_dsn'} = [ $dsn, $username, $password, $params ];
    return $self;
}

1;


lib/Prancer/Session/Store/Database/Driver/SQLite.pm
package Prancer::Session::Store::Database::Driver::SQLite;

use strict;
use warnings FATAL => 'all';

use version;
our $VERSION = '1.00';

use Prancer::Session::Store::Database::Driver;
use parent qw(Prancer::Session::Store::Database::Driver);

use Try::Tiny;
use Carp;

# even though this *should* work automatically, it was not
our @CARP_NOT = qw(Prancer Try::Tiny);

sub new {
    my $class = shift;

    try {
        require DBD::SQLite;
    } catch {
        my $error = (defined($_) ? $_ : "unknown");
        croak "could not initialize session handler: could not load DBD::SQLite: ${error}";
    };

    my $self = bless($class->SUPER::new(@_), $class);
    my $database  = $self->{'_database'};
    my $username  = $self->{'_username'};
    my $password  = $self->{'_password'};
    my $hostname  = $self->{'_hostname'};
    my $port      = $self->{'_port'};
    my $charset   = $self->{'_charset'};
    my $table     = $self->{'_table'};

    my $dsn = "dbi:SQLite:dbname=${database}";
    $dsn .= ";host=${hostname}" if defined($hostname);
    $dsn .= ";port=${port}" if defined($port);

    my $params = {
        'AutoCommit' => 0,
        'RaiseError' => 1,
        'PrintError' => 0,
    };
    if ($charset && $charset =~ /^utf8$/xi) {
        $params->{'sqlite_unicode'} = 1;
    }

    $self->{'_dsn'} = [ $dsn, $username, $password, $params ];
    return $self;
}

1;


lib/Prancer/Session/Store/Database/Driver.pm
package Prancer::Session::Store::Database::Driver;

use strict;
use warnings FATAL => 'all';

use version;
our $VERSION = '1.00';

use Plack::Session::Store;
use parent qw(Plack::Session::Store);

use DBI;
use Storable qw(nfreeze thaw);
use MIME::Base64 qw(encode_base64 decode_base64);
use Try::Tiny;
use Carp;

# even though this *should* work automatically, it was not
our @CARP_NOT = qw(Prancer Try::Tiny);

sub new {
    my ($class, $config) = @_;

    # this is the only required field
    unless ($config->{'database'}) {
        croak "could not initialize session handler: no database name configured";
    }

    # initialize the serializer that will be used
    my $self = bless($class->SUPER::new(%{$config || {}}), $class);
    $self->{'_serializer'}      = sub { encode_base64(nfreeze(shift)) };
    $self->{'_deserializer'}    = sub { thaw(decode_base64(shift)) };

    $self->{'_database'}        = $config->{'database'};
    $self->{'_username'}        = $config->{'username'};
    $self->{'_password'}        = $config->{'password'};
    $self->{'_hostname'}        = $config->{'hostname'};
    $self->{'_port'}            = $config->{'port'};
    $self->{'_autocommit'}      = $config->{'autocommit'};
    $self->{'_charset'}         = $config->{'charset'};
    $self->{'_check_threshold'} = $config->{'connection_check_threshold'} // 30;
    $self->{'_table'}           = $config->{'table'} || "sessions";
    $self->{'_timeout'}         = $config->{'expiration_timeout'} // 1800;
    $self->{'_autopurge'}       = $config->{'autopurge'} // 1;
    $self->{'_application'}     = $config->{'application'};

    # store a pool of database connection handles
    $self->{'_handles'} = {};

    return $self;
}

sub handle {
    my $self = shift;

    # to be fork safe and thread safe, use a combination of the PID and TID
    # (if running with use threads) to make sure no two processes/threads share
    # a handle. implementation based on DBIx::Connector by David E. Wheeler
    my $pid_tid = $$;
    $pid_tid .= "_" . threads->tid() if $INC{'threads.pm'};

    # see if we have a matching handle
    my $handle = $self->{'_handles'}->{$pid_tid} || undef;

    if ($handle->{'dbh'}) {
        if ($handle->{'dbh'}{'Active'} && $self->{'_check_threshold'} &&
            (time - $handle->{'last_connection_check'} < $self->{'_check_threshold'})) {

            # the handle has been checked recently so just return it
            return $handle->{'dbh'};
        } else {
            if ($self->_check_connection($handle->{'dbh'})) {
                $handle->{'last_connection_check'} = time;
                return $handle->{'dbh'};
            } else {
                # try to disconnect but don't care if it fails
                if ($handle->{'dbh'}) {
                    try { $handle->{'dbh'}->disconnect() } catch {};
                }

                # try to connect again and save the new handle
                $handle->{'dbh'} = $self->_get_connection();
                return $handle->{'dbh'};
            }
        }
    } else {
        $handle->{'dbh'} = $self->_get_connection();
        if ($handle->{'dbh'}) {
            $handle->{'last_connection_check'} = time;
            $self->{'_handles'}->{$pid_tid} = $handle;
            return $handle->{'dbh'};
        }
    }

    return;
}

sub _get_connection {
    my $self = shift;
    my $dbh = undef;

    try {
        $dbh = DBI->connect(@{$self->{'_dsn'}}) || die "${\$DBI::errstr}\n";
    } catch {
        my $error = (defined($_) ? $_ : "unknown");
        croak "could not initialize database connection '${\$self->{'_connection'}}': ${error}";
    };

    return $dbh;
}

# Check the connection is alive
sub _check_connection {
    my $self = shift;
    my $dbh = shift;
    return 0 unless $dbh;

    if ($dbh->{'Active'} && (my $result = $dbh->ping())) {
        if (int($result)) {
            # DB driver itself claims all is OK, trust it
            return 1;
        } else {
            # it was "0 but true", meaning the DBD doesn't implement ping and
            # instead we got the default DBI ping implementation. implement
            # our own basic check, by performing a real simple query.
            return try {
                return $dbh->do("SELECT 1");
            } catch {
                return 0;
            };
        }
    }

    return 0;
}

sub fetch {
    my ($self, $session_id) = @_;
    my $dbh = $self->handle();
    my $result = undef;

    try {
        my $now = time();
        my $table = $self->{'_table'};
        my $application = $self->{'_application'};

        my $sth = $dbh->prepare(qq|
            SELECT data
            FROM ${table}
            WHERE id = ?
              AND application = ?
              AND timeout >= ?
        |);
        $sth->execute($session_id, $application, $now);
        my ($data) = $sth->fetchrow();
        $sth->finish();

        # deserialize the data if there is any
        $result = (defined($data) ? $self->{'_deserializer'}->($data) : undef);

        $dbh->commit();
    } catch {
        try { $dbh->rollback(); } catch {};

        my $error = (defined($_) ? $_ : "unknown");
        carp "error fetching from session: ${error}";
    };

    return $result;
}

sub store {
    my ($self, $session_id, $data) = @_;
    my $dbh = $self->handle();

    try {
        my $now = time();
        my $table = $self->{'_table'};
        my $application = $self->{'_application'};
        my $timeout = ($now + $self->{'_timeout'});
        my $serialized = $self->{'_serializer'}->($data);

        my $insert_sth = $dbh->prepare(qq|
            INSERT INTO ${table} (id, application, timeout, data)
            SELECT ?, ?, ?, ?
            WHERE NOT EXISTS (
                SELECT 1
                FROM ${table}
                WHERE id = ?
                  AND application = ?
                  AND timeout >= ?
            )
        |);
        $insert_sth->execute($session_id, $application, $timeout, $serialized, $session_id, $application, $now);
        $insert_sth->finish();

        my $update_sth = $dbh->prepare(qq|
            UPDATE ${table}
            SET timeout = ?, data = ?
            WHERE id = ?
              AND application = ?
              AND timeout >= ?
        |);
        $update_sth->execute($timeout, $serialized, $session_id, $application, $now);
        $update_sth->finish();

        # 10% of the time we will also purge old sessions
        if ($self->{'_autopurge'}) {
            my $chance = rand();
            if ($chance <= 0.1) {
                my $delete_sth = $dbh->prepare(qq|
                    DELETE
                    FROM ${table}
                    WHERE application = ?
                      AND timeout < ?
                |);
                $delete_sth->execute($application, $now);
                $delete_sth->finish();
            }
        }

        $dbh->commit();
    } catch {
        try { $dbh->rollback(); } catch {};

        my $error = (defined($_) ? $_ : "unknown");
        carp "error fetching from session: ${error}";
    };

    return;
}

sub remove {
    my ($self, $session_id) = @_;
    my $dbh = $self->handle();

    try {
        my $table = $self->{'_table'};
        my $application = $self->{'_application'};

        my $sth = $dbh->prepare(qq|
            DELETE
            FROM ${table}
            WHERE id = ?
              AND application = ?
        |);
        $sth->execute($session_id, $application);
        $sth->finish();

        $dbh->commit();
    } catch {
        try { $dbh->rollback(); } catch {};

        my $error = (defined($_) ? $_ : "unknown");
        carp "error fetching from session: ${error}";
    };

    return;
}

1;
